File: 2023/day04/clojure/solution.clj
(ns solution
  (:require [clojure.string :as str]
            [clojure.set :as set]))

(def input
  (-> (slurp "../input.txt")
      str/trim
      (str/split #"\n")))

(defn parse-numbers [s]
  "Parse a space-separated string of numbers into a set"
  (->> (str/split (str/trim s) #"\s+")
       (map parse-long)
       set))

(defn parse-card [line]
  "Parse a card line into [winning-set have-set]"
  (let [[_ numbers] (str/split line #":")
        [winning-part have-part] (str/split numbers #"\|")]
    [(parse-numbers winning-part)
     (parse-numbers have-part)]))

(defn parse-cards []
  "Parse all cards from input"
  (map parse-card input))

(defn count-matches [[winning have]]
  "Count how many numbers match between winning and have"
  (count (set/intersection winning have)))

(defn card-score [card]
  "Calculate the score for a card: 2^(matches-1) if matches > 0, else 0"
  (let [matches (count-matches card)]
    (if (pos? matches)
      (bit-shift-left 1 (dec matches))
      0)))

(defn part1 []
  (->> (parse-cards)
       (map card-score)
       (reduce + 0)))

(defn part2 []
  (let [cards (vec (parse-cards))
        n (count cards)
        matches (mapv count-matches cards)
        initial-copies (vec (repeat n 1))]
    (->> (range n)
         (reduce (fn [copies i]
                   (let [m (nth matches i)
                         card-copies (nth copies i)]
                     (reduce (fn [c j]
                               (update c j + card-copies))
                             copies
                             (range (inc i) (min (+ i 1 m) n)))))
                 initial-copies)
         (reduce + 0))))

;; Main execution
(println "Part 1:" (part1))
(println "Part 2:" (part2))

File: 2023/day04/php/solution.php
<?php
declare(strict_types=1);

/**
 * Parse a space-separated string of numbers into an associative array.
 * Uses keys for O(1) lookup during intersection counting.
 *
 * @param string $str Space-separated numbers
 * @return array<int, bool> Numbers as keys with true values
 */
function parseNumbers(string $str): array
{
    $result = [];
    foreach (preg_split('/\s+/', trim($str)) as $num) {
        if ($num !== '') {
            $result[(int)$num] = true;
        }
    }
    return $result;
}

/**
 * Parse all scratchcard lines into structured card data.
 *
 * @param array<string> $lines Input lines
 * @return array<array{0: array<int, bool>, 1: array<int, bool>}> Parsed cards
 */
function parseCards(array $lines): array
{
    $cards = [];
    foreach ($lines as $line) {
        [, $numbers] = explode(':', $line);
        [$winningPart, $havePart] = explode('|', $numbers);

        $cards[] = [
            parseNumbers($winningPart),
            parseNumbers($havePart),
        ];
    }
    return $cards;
}

/**
 * Count matching numbers between winning numbers and numbers we have.
 *
 * @param array<int, bool> $winning Winning numbers
 * @param array<int, bool> $have Numbers we have
 * @return int Number of matches
 */
function countMatches(array $winning, array $have): int
{
    return count(array_intersect_key($have, $winning));
}

/**
 * Part 1: Calculate total points from scratchcards.
 * Each card scores 2^(matches-1) points if it has any matches.
 *
 * @param array<array{0: array<int, bool>, 1: array<int, bool>}> $cards Parsed cards
 * @return int Total points
 */
function part1(array $cards): int
{
    $total = 0;

    foreach ($cards as [$winning, $have]) {
        $matches = countMatches($winning, $have);
        if ($matches > 0) {
            $total += 2 ** ($matches - 1);
        }
    }

    return $total;
}

/**
 * Part 2: Calculate total number of scratchcards including copies.
 * Each match on card N wins a copy of cards N+1 through N+matches.
 *
 * @param array<array{0: array<int, bool>, 1: array<int, bool>}> $cards Parsed cards
 * @return int Total number of cards
 */
function part2(array $cards): int
{
    $numCards = count($cards);

    // Pre-calculate matches for each card
    $matchCounts = array_map(
        fn(array $card): int => countMatches($card[0], $card[1]),
        $cards
    );

    // Track copies of each card (start with 1 original each)
    $copies = array_fill(0, $numCards, 1);

    // Process each card and add copies to subsequent cards
    for ($i = 0; $i < $numCards; $i++) {
        $matches = $matchCounts[$i];
        for ($j = $i + 1; $j < min($i + 1 + $matches, $numCards); $j++) {
            $copies[$j] += $copies[$i];
        }
    }

    return array_sum($copies);
}

// Read input
$input = file_get_contents(__DIR__ . '/../input.txt');
$lines = explode("\n", trim($input));

// Parse cards once and pass to both parts
$cards = parseCards($lines);

echo "Part 1: " . part1($cards) . "\n";
echo "Part 2: " . part2($cards) . "\n";
